x-pack/plugins/spaces/server/lib/space_request_interceptors.js

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { wrapError } from './errors';
+
+export function initSpacesRequestInterceptors(server) {
+  server.ext('onRequest', async function spacesOnRequestHandler(request, reply) {
+    const path = request.path;
+
+    // If navigating within the context of a space, then we store the Space's URL Cpntext on the request,
+    // and rewrire the request to not include the space identifier in the URL.
+
+    if (path.startsWith('/s/')) {
+      const pathParts = path.split('/');
+
+      const spaceUrlContext = pathParts[2];
+
+      const reqBasePath = `/s/${spaceUrlContext}`;
+      request.setBasePath(reqBasePath);
+
+      const newUrl = {
+        ...request.url,
+        path: path.substr(reqBasePath.length) || '/',
+        pathname: path.substr(reqBasePath.length) || '/',
+        href: path.substr(reqBasePath.length) || '/'
+      };
+
+      request.setUrl(newUrl);
+      request.app._spaceUrlContext = spaceUrlContext;
```

Comment:
Are we only using `_spaceUrlContext` below? I've seen usages of Weakmap to do similar things without "leaking" the variable, perhaps we could do something similar here?

Reply:
I wonder if we could expose the UrlContext on the request itself, that way it's more generic?

Reply:
hmm, we certainly could. 
For the Saved Objects Client, I imagine that we'll be augmenting the reads/writes to include the new `spaceId` property on the index mapping, so we will need a way to lookup the `spaceId` from the URL Context. I'm worried that decoupling the URL Context from Spaces too much will make that translation confusing to follow.

I also thought about just storing the `spaceId` instead, but that introduces a lot of unnecessary overhead on each request (because we have to look it up), even if the space id isn't needed for it.